### PR TITLE
chore: Document name resolution

### PR DESCRIPTION
**Pull request by jfecher**
# Related issue(s)

(None)

# Description

## Summary of changes

Adds documentation to various items in the name resolution pass.

Documents some higher-level architecture of the pass, most notably that it has two phases: the definition collection phase, followed by the actual name resolution phase.

### CODE REVIEW

File: crates/noirc_frontend/src/hir_def/types.rs [hunk 1]
```diff
@@ -222,16 +241,31 @@ impl TypeBinding {
...
+/// polymorphically comptime because they can be used in both contexts.
 #[derive(Debug, Clone, Eq)]
 pub enum CompTime {
     // Yes and No variants have optional spans representing the location in the source code
     // which caused them to be compile time.
     Yes(Option<Span>),
     No(Option<Span>),
-    Maybe(TypeVariableId, Rc<RefCell<Option<CompTime>>>),
+
+    /// Maybe has an id and shared inner rebindable reference. Similar to type variables rebinding
```
Comment (kevaundray) [inline crates/noirc_frontend/src/hir_def/types.rs:266]: rebindable does not seem to be a word, can you rephrase?

Comment (jfecher) [reply crates/noirc_frontend/src/hir_def/types.rs:266]: Seems like another rare term the spell checker is unaware of (e.g. it is also used by haskell: https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/rebindable_syntax.html). "shared inner reference that can be rebound" perhaps.

**Comment by TomAFrench** [comment]
> LGTM. Weird that clippy has now started to fail

We've started enforcing clippy on all crates in the workspace as previously it was only enforcing it on `nargo`. This warning is fixed in https://github.com/noir-lang/noir/pull/918 so we just need to merge master back into this branch.
